RenderController : Don't dirty all locations when non-attribute globals change

themissingcow opened the pull request:
As `m_fullAttributes` also contained `sets`, our comparison in `updateAttributes` would always fail and consequently dirty the entire tree for any non-attribute global. Badness!

Fixes
-----

- RenderController : Fixed bug that caused unnecessary attribute updates when changing globals.

themissingcow commented on src/GafferScene/RenderController.cpp:
```diff
@@ -95,6 +97,61 @@ bool cameraGlobalsChanged( const CompoundObject *globals, const CompoundObject *
+// that are only present in previousFullAttributes.
+bool globalAttributesChanged( const CompoundObject* globalAttributes, const CompoundObject* previousFullAttributes )
+{
+	static std::set<IECore::InternedString> ignoredMembers = { IECore::InternedString( "sets" ) };
```
Went back and forwards on just hard-coding `if( it1->first != staticSetsString )` and the offsets, but it seemed more fragile, given the use in the early-out clauses, etc... Happy to do that instead if preferred.

johnhaddon replied:
"Why you shouldn't use set (and what you should use instead)" is a decent read : http://lafstern.org/matt/col1.pdf. The relevant bits here are that our number of elements is small and we build the container once at startup and never need to edit it again, hence `set` has overheads we pay for but don't benefit from. I agree the code maintenance is a little easier with `ignoredMembers` though, so suggest using `const boost::flat_set`instead (note the `const`).

themissingcow replied:
Ah, good to know, thanks for the link. https://github.com/GafferHQ/gaffer/pull/3606/commits/8e9bcd830907ce70768a570f0b9e2a4c97cdfc9b.

johnhaddon commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -372,5 +372,130 @@ def testHideLinkedLight( self ) :
 		self.assertIsNone( renderer.capturedObject( "/group/defaultLight" ) )
 		self.assertEqual( capturedPlane.capturedLinks( "lights" ), set() )
 
+	def testAttributeDirtyPropogation( self ) :
```
Propogation -> Propagation

themissingcow replied:
Doh! Sorry. https://github.com/GafferHQ/gaffer/pull/3606/commits/9676acb27e0c3cf4a55e025abc8816ebd334bdea

themissingcow commented:
> I don't see anything especially risky here, but wonder what you think to targeting 55 (or even 56) rather than 54?

@johnhaddon I've addressed the review comments. Let me know if you're happy, and I'll move to `0.55_maintenance` when I squish.

johnhaddon commented:
> Let me know if you're happy, and I'll move to 0.55_maintenance when I squish.

Thanks Tom - LGTM.

themissingcow commented:
@johnhaddon squished and moved to `0.55_maintenance` should be good when CI has gone through.
